### PR TITLE
fix(BS-15447) Invalid JSON produced by /API/bpm/activityVariable when process variable is null

### PR DESCRIPTION
**Pull request by abirembaut**
- Handle the case where the variable value is null
- add unit test

Covers [BS-15447](https://bonitasoft.atlassian.net/browse/BS-15447)

### CODE REVIEW

File: server/src/test/java/org/bonitasoft/web/rest/server/api/bpm/flownode/ActivityVariableResourceTest.java [hunk 1]
```diff
@@ -91,36 +91,35 @@ public void should_return() throws DataNotFoundException {
         assertThat(dataInstanceResult).isEqualTo(dataInstance);
     }
 
-    private DataInstanceImpl createShortTextDataInstance(String value) {
-        DataDefinition dataDefinition = createDataDefinition();
+    private DataInstanceImpl createShortTextDataInstance(final String value) {
```
Comment (velcrin) [inline server/src/test/java/org/bonitasoft/web/rest/server/api/bpm/flownode/ActivityVariableResourceTest.java]: Are all those `final` necessary? Thought I heard they were out of the formatting template.

File: server/src/main/java/org/bonitasoft/web/rest/server/utils/JacksonSerializerHelper.java [hunk 2]
```diff
@@ -36,21 +36,14 @@ public JacksonSerializerHelper() {
         numberTypes.add(Double.class.getCanonicalName());
```
Comment (velcrin) [inline server/src/main/java/org/bonitasoft/web/rest/server/utils/JacksonSerializerHelper.java]: Could this not be a status list?

File: server/src/main/java/org/bonitasoft/web/rest/server/utils/JacksonSerializerHelper.java [hunk 3]
```diff
@@ -36,21 +36,14 @@ public JacksonSerializerHelper() {
...
 
-    protected String getStringValue(Serializable value) {
-        if (value == null) {
-            return null;
-        }
-        return String.valueOf(value);
-    }
-
-    protected void writeValueAndStringValue(JsonGenerator jgen, String fieldName, Serializable value) throws IOException {
+    protected void writeValueAndStringValue(final JsonGenerator jgen, final String fieldName, final Serializable value) throws IOException {
```
Comment (velcrin) [inline server/src/main/java/org/bonitasoft/web/rest/server/utils/JacksonSerializerHelper.java]: Why is this method protected? Also I think the name `writeNumberField` would be more accurate.

File: server/src/main/java/org/bonitasoft/web/rest/server/utils/JacksonSerializerHelper.java [hunk 4]
```diff
@@ -36,21 +36,14 @@ public JacksonSerializerHelper() {
...
         jgen.writeObjectField(fieldName, value);
-        if (numberTypes.contains(value.getClass().getCanonicalName())) {
-            jgen.writeObjectField(fieldName + "_string", getStringValue(value));
+        if (value != null && numberTypes.contains(value.getClass().getCanonicalName())) {
```
Comment (velcrin) [inline server/src/main/java/org/bonitasoft/web/rest/server/utils/JacksonSerializerHelper.java]: `numberTypes` should more exhaustive like `supportedNumberTypes`

File: server/src/main/java/org/bonitasoft/web/rest/server/utils/JacksonSerializerHelper.java (hunk 2 above)
Comment (velcrin) [inline server/src/main/java/org/bonitasoft/web/rest/server/utils/JacksonSerializerHelper.java]: Also should the name of this class not be `DataInstanceSerializer`?

**Comment by velcrin** [comment]
The fix looks good but I did some comments on the code itself.
I know most of this is not your own but well lets apply the boyscout rule!

File: server/src/main/java/org/bonitasoft/web/rest/server/utils/JacksonSerializerHelper.java [hunk 5]
```diff
@@ -36,21 +36,14 @@ public JacksonSerializerHelper() {
...
+            jgen.writeObjectField(fieldName + "_string", String.valueOf(value));
         }
     }
 
-    public void serializeDataInstance(DataInstanceImpl value, JsonGenerator jgen) throws IOException {
+    public void serializeDataInstance(final DataInstanceImpl value, final JsonGenerator jgen) throws IOException {
```
Comment (velcrin) [inline server/src/main/java/org/bonitasoft/web/rest/server/utils/JacksonSerializerHelper.java]: We could rename this method `serialize` if the class was properly named `DataInstanceSerializer`

File: server/src/test/java/org/bonitasoft/web/rest/server/api/bpm/flownode/ActivityVariableResourceTest.java (hunk 1 above)
Comment (abirembaut) [inline server/src/test/java/org/bonitasoft/web/rest/server/api/bpm/flownode/ActivityVariableResourceTest.java]: Not really in this particular case.
I will remove them.
But if there was really a decision on this, the templates haven't been updated : https://github.com/bonitasoft/bonita-internal-tools/tree/master/codestyle

File: server/src/main/java/org/bonitasoft/web/rest/server/utils/JacksonSerializerHelper.java (hunk 4 above)
Comment (abirembaut) [inline server/src/main/java/org/bonitasoft/web/rest/server/utils/JacksonSerializerHelper.java]: I agree. having the attribute always present makes the API easier to use, I think.

**Comment by abirembaut** [comment]
Good points !
I updated the PR accordingly

**Pull request merged**